warehouse/accounts/services.py

```diff
@@ -631,6 +633,71 @@ def get_password_timestamp(self, user_id):
         user = self.get_user(user_id)
         return user.password_date.timestamp() if user.password_date is not None else 0
 
+    def needs_tos_update(
+        self, user_id, revision, ignore_flashed=False, ignore_notified=False
+    ):
+        query = self.db.query(UserTermsOfServiceEngagement).filter(
+            UserTermsOfServiceEngagement.user_id == user_id,
+            UserTermsOfServiceEngagement.revision == revision,
+        )
+        engagements = [
+            TermsOfServiceEngagement.Viewed,
+            TermsOfServiceEngagement.Agreed,
+        ]
+        if not ignore_flashed:
+            engagements.append(TermsOfServiceEngagement.Flashed)
+        if not ignore_notified:
+            engagements.append(TermsOfServiceEngagement.Notified)
+
+        first_engagement_all = (
+            query.filter(
+                UserTermsOfServiceEngagement.datetime
+                < datetime.datetime.now(datetime.UTC) - datetime.timedelta(days=30)
+            )
+            .order_by(
+                UserTermsOfServiceEngagement.datetime,
+            )
+            .first()
+        )
+        if first_engagement_all is not None:
+            return False
+
+        first_engagement = query.filter(
+            UserTermsOfServiceEngagement.engagement.in_(engagements)
+        ).first()
+        if first_engagement is None:
+            return True
+
+        return False
```

Comment:
I'm finding this confusing to say in prose.

> `first_engagement_all` will contain any engagements made by <this user> for <this revision> made over 30 days ago.
> If there are any older than 30 days, then `needs_tos_update` will be `False`
> During login we don't need to flash them (the only place this function is currently called)

> `first_engagement` will contain the first of any applicable engagements, based on the function inputs (currently always True, so all 4 are checked)
> If there's none, and return `True`
> During login, flash them the ToS notice

Is that what you intended? Variable naming might help here, as well as inline comments.

Reply:
renamed, commented, and refactored a bit in f3929d800